src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs

```diff
@@ -0,0 +1,311 @@
+                            newItem.Properties["related"] = (string)related;
+                        }
+                        additionalAction?.Invoke(newItem);
+                        yield return newItem;
+                    }
+                    // Take only the first group that has items
+                    break;
+                }
+            }
+
+            yield break;
+        }
+
+        private static async Task<NuGetVersion> DownloadAndExtractPackage(
+            PackageLocation packageLocation,
+            PackageId packageId,
+            INuGetPackageDownloader _nugetPackageDownloader,
+            string hashPathLocation,
+            IToolPackageStore toolPackageStore
+            )
+        {
+            var packageSourceLocation = new PackageSourceLocation(packageLocation.NugetConfig, packageLocation.RootConfigDirectory, null, packageLocation.AdditionalFeeds);
+            var packagePath = await _nugetPackageDownloader.DownloadPackageAsync(packageId, null, packageSourceLocation);
+
+            // look for package on disk and read the version
+            NuGetVersion version;
+
+            using (FileStream packageStream = File.OpenRead(packagePath))
+            {
+                PackageArchiveReader reader = new PackageArchiveReader(packageStream);
+                version = new NuspecReader(reader.GetNuspec()).GetVersion();
+
+                var packageDirectory = toolPackageStore.GetPackageDirectory(packageId, version);
+
+                if (Directory.Exists(packageDirectory.Value))
+                {
+                    throw new ToolPackageException(
+                        string.Format(
+                            CommonLocalizableStrings.ToolPackageConflictPackageId,
+                            packageId,
+                            version.ToNormalizedString()));
+                }
+
+                if (Directory.Exists(packagePath))
+                {
+                    throw new ToolPackageException(
+                        string.Format(
+                            CommonLocalizableStrings.ToolPackageConflictPackageId,
+                            packageId,
+                            version.ToNormalizedString()));
+                }
+
+                var packageHash = Convert.ToBase64String(new CryptoHashProvider("SHA512").CalculateHash(reader.GetNuspec()));
+                var hashPath = new VersionFolderPathResolver(hashPathLocation).GetHashPath(packageId.ToString(), version);
+
+                Directory.CreateDirectory(Path.GetDirectoryName(hashPath));
+                File.WriteAllText(hashPath, packageHash);
+            }
+
+            // Extract the package
+            var nupkgDir = Path.Combine(hashPathLocation, packageId.ToString(), version.ToString());
+            var filesInPackage = await _nugetPackageDownloader.ExtractPackageAsync(packagePath, new DirectoryPath(nupkgDir));
+
+            return version;
+        }
+
+        private static void CreateAssetFiles(
```

Comment:
I'm sure @zivkan brought it up somewhere, but creating an assets file is probably not the best long term approach. Since you're already doing a custom installation, are there are plans to move away from the assets files?

Reply:
https://github.com/dotnet/sdk/pull/33835#discussion_r1282977739

Yes, if the only file that matters is `DotnetToolSettings.xml`, then that can be found with much less complexity via `Directory.EnumerateFiles`.

Reply:
to be follow up in #35283